core/src/main/java/org/infinispan/util/concurrent/BlockingHandler.java

```diff
@@ -0,0 +1,158 @@
+package org.infinispan.util.concurrent;
+
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.reactivestreams.Publisher;
+
+/**
+ * A handler to ensure that blocking operations are performed in the correct manner. That is that a blocking operation
```

Comment:
The first sentence mentions "blocking operation" and "correct manner" but they're just names without any meaning at this point. I'd say "blocking task", because "task" is more familiar when coupled with "executor", maybe

```
Runs tasks on the blocking executor and returns a {@code CompletionStage} or {@code Publisher} that continues on the non-blocking executor, similar to {@code stage.handleAsync(callback, blockingExecutor).whenCompleteAsync(NOOP, nonBlockingExecutor)}.
<p>
If the current thread is blocking, it blocks until the task may run, runs the task in the current thread, and returns a completed {@code CompletionStage}, so it <em>does not</em> continue the execution on the non-blocking executor.
```

Reply:
Sure, I can actually add some more details about this.